src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java

```diff
@@ -0,0 +1,3516 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.util.random;
+
+import java.lang.annotation.*;
+import java.math.BigInteger;
+import java.util.function.Consumer;
+import java.util.function.DoubleConsumer;
+import java.util.function.IntConsumer;
+import java.util.function.LongConsumer;
+import java.util.random.RandomGenerator;
+import java.util.random.RandomGenerator.SplittableGenerator;
+import java.util.Spliterator;
+import java.util.stream.DoubleStream;
+import java.util.stream.IntStream;
+import java.util.stream.LongStream;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+/**
+ * Low-level utility methods helpful for implementing (pseudo)random number
+ * generators.
+ *
+ * <p> This class is mostly for library writers creating specific
+ * implementations of the interface {@link RandomGenerator}. As an
+ * internal package it is not intended for general use.
+ *
+ * @since 17
+ * @hidden
+ *
+ */
+public class RandomSupport {
+    /**
+     * Annotation providing RandomGenerator properties.
+     */
+    @Retention(RetentionPolicy.RUNTIME)
+    @Target(ElementType.TYPE)
+    public @interface RandomGeneratorProperties {
+        /**
+         * Name of algorithm.
+         */
+        String name();
+
+        /**
+         * Category of algorithm.
+         */
+        String group() default "Legacy";
+
+        /**
+         * Algorithm period defined as:
+         *
+         * BigInteger.ONE.shiftLeft(i)
+         *               .subtract(j)
+         *               .shiftLeft(k)
+         */
+        int i() default 0;
+        int j() default 0;
+        int k() default 0;
+
+        /**
+         * The equidistribution of the algorithm.
+         */
+        int equidistribution() default Integer.MAX_VALUE;
+
+        /**
+         * Is the algorithm based on entropy (true random.)
+         */
+        boolean isStochastic() default false;
+
+        /**
+         * Is the algorithm assisted by hardware (fast true random.)
+         */
+        boolean isHardware() default false;
+    }
+
+    /*
+     * Implementation Overview.
+     *
+     * This class provides utility methods and constants frequently
+     * useful in the implementation of (pseudo)random number generators
+     * that satisfy the interface {@link RandomGenerator}.
+     *
+     * File organization: First some message strings, then the main
+     * public methods, followed by a non-public base spliterator class.
+     */
+
+    // IllegalArgumentException messages
+    static final String BAD_SIZE = "size must be non-negative";
+    static final String BAD_DISTANCE = "jump distance must be finite, positive, and an exact integer";
+    static final String BAD_BOUND = "bound must be positive";
+    static final String BAD_FLOATING_BOUND = "bound must be finite and positive";
+    static final String BAD_RANGE = "bound must be greater than origin";
```

Comment:
Unused fields in Random class now can be cleaned up: `java.util.Random#BadRange`, `java.util.Random#BadSize`

Reply:
Good